tsinfer/inference.py

```diff
@@ -890,23 +980,77 @@ def __init__(
             [inference_site_position, [sample_data.sequence_length]]
         )
         self.position_map[0] = 0
+        self.recombination = np.zeros(self.num_sites)  # TODO: reduce len by 1
+        self.mismatch = np.zeros(self.num_sites)
+
+        # Set the canonical 'recombination' and 'mismatch' arrays to appropriate values
+        # NB 'recombination' must be of length num_sites-1, mismatch of length num_sites
+
+        # First set defaults if None provided
+        if recombination_rate is None and recombination is None:
+            if mismatch_ratio is None and mismatch is None:
+                # Special case: revert to tsinfer 0.1 behaviour with no mismatch allowed
+                mismatch = np.full(self.num_sites, 1e-20)  # A miniscule value
+                # arbitrary, large relative to mismatch, but < 1
+                if self.num_sites == 0:
+                    recombination = []
+                else:
+                    recombination = np.full(self.num_sites - 1, 1e-2)
+                # Shouldn't need insane precision, although 3 is not enough in trials
```

Comment:
Let's keep precision out of this change - there's enough moving parts here. Keep a constant default, we can use something more sophisticated later.